### PR TITLE
Avoid removal of non-expired PIDs when log cleaning

### PR DESCRIPTION
**Pull request by hachikuji**

### CODE REVIEW

**Comment by apurvam** [comment]
It looks like this needs another rebase :(

File: core/src/main/scala/kafka/log/ProducerIdMapping.scala [hunk 1]
```diff
@@ -162,6 +161,8 @@ class ProducerIdMapping(val topicPartition: TopicPartition,
    */
   def mapEndOffset = lastMapOffset
 
+  def activePids: immutable.Map[Long, PidEntry] = pidMap.toMap
```
Comment (guozhangwang) [inline core/src/main/scala/kafka/log/ProducerIdMapping.scala]: I think we can just inline this function inside `Log.scala`.

File: core/src/main/scala/kafka/log/ProducerIdMapping.scala [hunk 2]
```diff
@@ -147,7 +145,8 @@ object ProducerIdMapping {
 @nonthreadsafe
 class ProducerIdMapping(val topicPartition: TopicPartition,
                         val config: LogConfig,
-                        val snapParentDir: File) extends Logging {
+                        val snapParentDir: File,
```
Comment (guozhangwang) [inline core/src/main/scala/kafka/log/ProducerIdMapping.scala]: Are these two lines part of this PR?

File: clients/src/main/java/org/apache/kafka/common/record/MemoryRecords.java [hunk 3]
```diff
@@ -331,13 +331,26 @@ public static MemoryRecords withRecords(long initialOffset, CompressionType comp
         return withRecords(LogEntry.CURRENT_MAGIC_VALUE, initialOffset, compressionType, records);
     }
 
+    public static MemoryRecords withRecords(long initialOffset, CompressionType compressionType, Long pid,
```
Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/common/record/MemoryRecords.java]: Are these changes part of this PR?

Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/common/record/MemoryRecords.java]: Yes, I needed this in the test case below.

File: core/src/main/scala/kafka/log/ProducerIdMapping.scala (hunk 2 above)
Comment (hachikuji) [inline core/src/main/scala/kafka/log/ProducerIdMapping.scala]: Ack. Probably thought I would need this initially, but ended up doing that part in a separate PR. I'll remove it here.

File: core/src/main/scala/kafka/log/ProducerIdMapping.scala (hunk 1 above)
Comment (hachikuji) [inline core/src/main/scala/kafka/log/ProducerIdMapping.scala]: Seems nicer to keep the mutability of the internal map hidden from `Log`.

**Pull request merged**